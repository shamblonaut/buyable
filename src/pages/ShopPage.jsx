import { useEffect, useState } from "react";
import { useOutletContext, useSearchParams } from "react-router-dom";
import styled from "styled-components";

import SearchBar from "../components/SearchBar";
import ProductCard from "../components/ProductCard.jsx";

const Page = styled.main`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

const ProductList = styled.ul`
  list-style: none;
  padding: 16px;
  width: 100%;
  display: grid;
  grid-template-columns: repeat(auto-fit, 180px);
  gap: 16px;
  margin-bottom: 64px;
`;

const Info = styled.p`
  text-align: center;
`;

const ShopPage = () => {
  const [searchParams, setSearchParams] = useSearchParams();

  const [searchQuery, setSearchQuery] = useState(
    searchParams.get("query") ?? "",
  );
  const [filteredProducts, setFilteredProducts] = useState([]);

  const { productsData, productsError, productsLoading } = useOutletContext();

  useEffect(() => {
    if (!searchQuery && productsData) {
      setFilteredProducts(productsData);
      setSearchParams({});
      return;
    } else if (productsData) {
      setFilteredProducts(
        productsData.filter(
          (product) =>
            searchQuery === "" ||
            product.title.toLowerCase().includes(searchQuery.toLowerCase()) ||
            product.category.includes(searchQuery.toLowerCase()),
        ),
      );
      setSearchParams({ query: searchQuery });
    }
  }, [searchQuery, productsData, setSearchParams]);

  return (
    <Page>
      <SearchBar
        prompt="Search Product"
        query={searchQuery}
        setQuery={setSearchQuery}
      />
      {productsLoading ? (
        <Info>Loading Products...</Info>
      ) : productsError ? (
        <Info>
          Error while fetching products:
          <br />
          {productsError}
        </Info>
      ) : filteredProducts.length > 0 ? (
        <ProductList>
          {filteredProducts.map((product) => (
            <li key={product.id}>
              <ProductCard product={product} />
            </li>
          ))}
        </ProductList>
      ) : (
        searchQuery && <Info>No products found for "{searchQuery}"</Info>
      )}
    </Page>
  );
};

export default ShopPage;
