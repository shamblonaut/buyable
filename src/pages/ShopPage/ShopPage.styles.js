import styled from "styled-components";

import { Info as CommonInfo } from "@/styles";

export const Info = styled(CommonInfo)`
  flex: 0.5;
  max-width: 90%;
  font-size: 1.5rem;
`;
