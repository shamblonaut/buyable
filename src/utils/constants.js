export const AppPosition = Object.freeze({
  HOME: 0,
  SHOP: 1,
  CART: 2,
});
