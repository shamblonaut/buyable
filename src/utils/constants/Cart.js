export const CartAction = Object.freeze({
  ADDED: "added_to_cart",
  REMOVED: "removed_from_cart",
  INCREMENTED: "incremented_cart_quantity",
  DECREMENTED: "decremented_cart_quantity",
  EDITED: "edited_cart_quantity",
  CLEARED: "cleared_cart",
});
