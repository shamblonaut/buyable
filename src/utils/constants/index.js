import { AppPosition } from "./App";

export { AppPosition };
