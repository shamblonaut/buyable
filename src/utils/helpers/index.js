import { getPositionRoute } from "./App";
import { getRatingArray } from "./Rating";
import { toTitleCase } from "./Misc";

export { getPositionRoute, getRatingArray, toTitleCase };
