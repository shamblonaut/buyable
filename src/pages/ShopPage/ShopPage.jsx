import { useEffect } from "react";
import { useOutletContext } from "react-router-dom";

import { AppPosition } from "@/utils/constants";

import { ProductList } from "@/components";
import { Page } from "@/styles";
import { Info } from "./ShopPage.styles";

const ShopPage = () => {
  const { productsData, cart, setCart, setAppPosition } = useOutletContext();

  useEffect(() => setAppPosition(AppPosition.SHOP), [setAppPosition]);

  return (
    <Page>
      {productsData.loading ? (
        <Info>Loading products...</Info>
      ) : productsData.error || !productsData.data ? (
        <Info>
          Could not load products
          {productsData.error?.message && (
            <>
              :
              <br />
              {productsData.error.message}
            </>
          )}
        </Info>
      ) : (
        productsData.data && (
          <ProductList
            products={productsData.data}
            cart={cart}
            setCart={setCart}
          />
        )
      )}
    </Page>
  );
};

export default ShopPage;
