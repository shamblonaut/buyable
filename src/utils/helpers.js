import { AppPosition } from "./constants";

export const getRatingArray = (rating) => {
  if (rating > 5.0) {
    throw new Error("Invalid rating");
  }

  let ratingArray = [];
  for (let i = 0; i < 5; i++) {
    const remainder = rating - i;
    ratingArray.push(
      remainder > 1 ? 1 : remainder > 0 ? Number(remainder.toFixed(1)) : 0,
    );
  }

  return ratingArray;
};

export const getPositionRoute = (appPosition) => {
  switch (appPosition) {
    case AppPosition.HOME:
      return "/";
    case AppPosition.SHOP:
      return "/shop";
    case AppPosition.CART:
      return "/cart";
  }
};

export const toTitleCase = (string) => {
  let result = "";
  for (const word of string.split(" ")) {
    result += word[0].toUpperCase() + word.substring(1) + " ";
  }
  return result.trim();
};
