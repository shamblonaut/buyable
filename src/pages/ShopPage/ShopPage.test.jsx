import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen } from "@testing-library/react";

import { AppPosition } from "@/utils/constants";

import { mockProductList, mockCart } from "@tests/data";
import OutletWrapper from "@tests/components/OutletWrapper";

import ShopPage from "./ShopPage";

const mockOutletContext = {
  productsData: {
    data: mockProductList,
    error: null,
    loading: false,
  },
  cart: mockCart,
  setCart: vi.spyOn(
    {
      setCart: vi.fn(),
    },
    "setCart",
  ),
  setAppPosition: vi.spyOn(
    {
      setAppPosition: vi.fn(),
    },
    "setAppPosition",
  ),
};

vi.mock("@/components", async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    ProductList: ({ products }) => (
      <ul>
        {products.map((product) => (
          <li data-testid="product" key={product.id}>
            {product.id}
          </li>
        ))}
      </ul>
    ),
  };
});

describe("Shop Page", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders correctly", () => {
    const { container } = render(
      <OutletWrapper context={mockOutletContext}>
        <ShopPage />
      </OutletWrapper>,
    );

    expect(container).toMatchSnapshot();
  });

  it("sets the app position correctly on load", () => {
    render(
      <OutletWrapper context={mockOutletContext}>
        <ShopPage />
      </OutletWrapper>,
    );

    expect(mockOutletContext.setAppPosition).toHaveBeenCalledWith(
      AppPosition.SHOP,
    );
  });

  it("shows loading text when data is loading", () => {
    render(
      <OutletWrapper
        context={{
          ...mockOutletContext,
          productsData: {
            data: null,
            error: null,
            loading: true,
          },
        }}
      >
        <ShopPage />
      </OutletWrapper>,
    );

    expect(
      screen.getByText("loading products", { exact: false }),
    ).toBeInTheDocument();
  });

  it("shows error text when there is an error in loading the data", () => {
    render(
      <OutletWrapper
        context={{
          ...mockOutletContext,
          productsData: {
            data: null,
            error: new Error("Mock Error"),
            loading: false,
          },
        }}
      >
        <ShopPage />
      </OutletWrapper>,
    );

    expect(
      screen.getByText("could not load products", { exact: false }),
    ).toBeInTheDocument();
    expect(
      screen.getByText("mock error", { exact: false }),
    ).toBeInTheDocument();
  });

  it("shows the correct products in product list", () => {
    render(
      <OutletWrapper context={mockOutletContext}>
        <ShopPage />
      </OutletWrapper>,
    );

    screen.getAllByTestId("product").forEach((productItem, index) => {
      expect(productItem.textContent).toBe(
        mockProductList[index].id.toString(),
      );
    });
  });
});
