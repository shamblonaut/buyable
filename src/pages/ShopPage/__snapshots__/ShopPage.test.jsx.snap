// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shop Page > renders correctly 1`] = `
<div>
  <main
    class="sc-hwkwBN kikxJF"
  >
    <ul>
      <li
        data-testid="product"
      >
        1
      </li>
      <li
        data-testid="product"
      >
        2
      </li>
      <li
        data-testid="product"
      >
        3
      </li>
      <li
        data-testid="product"
      >
        4
      </li>
      <li
        data-testid="product"
      >
        5
      </li>
      <li
        data-testid="product"
      >
        6
      </li>
      <li
        data-testid="product"
      >
        7
      </li>
      <li
        data-testid="product"
      >
        8
      </li>
      <li
        data-testid="product"
      >
        9
      </li>
      <li
        data-testid="product"
      >
        10
      </li>
    </ul>
  </main>
</div>
`;
